update-format-crau/src/bin/test.rs

```diff
@@ -79,12 +83,91 @@ fn main() -> Result<(), Box<dyn Error>> {
         _ => None,
     };
 
-    let signatures = match signatures_bytes {
-        Some(ref bytes) => Some(proto::Signatures::parse_from_bytes(bytes)?),
-        None => None,
+    Ok(signatures_bytes.unwrap())
+}
+
+// parse_signature_data takes SigBuffer and public key file path.
+// Return only actual data, without version and special fields.
+fn parse_signature_data(sigbytes: &[u8], pubkeyfile: &str) -> Result<Box<[u8]>, Box<dyn Error>> {
+    // Signatures has a container of the fields, i.e. version, data, and
+    // special fields.
+    let sigmessage = proto::Signatures::parse_from_bytes(sigbytes)?;
+    let siglen = sigbytes.len();
+
+    let mut sigbs: Box<[u8]> = vec![0; siglen].into_boxed_slice();
+    let mut sigfound: bool = false;
+
+    // sigmessages.signatures[] has a single element in case of dev update payloads,
+    // while it could have multiple elements in case of production update payloads.
+    // For now we assume only dev update payloads are supported.
+    // Return the first valid signature, iterate into the next slot if error.
+    for sigx in &sigmessage.signatures {
+        match verify_sig_pubkey(sigx, pubkeyfile) {
+            Ok(sigxvec) => {
+                sigbs = sigxvec.into_boxed_slice();
+                sigfound = true;
+                break;
+            }
+            Err(_) => {
+                continue;
+            }
+        };
+    }
+
+    if !sigfound {
+        return Err("No signature found".into());
+    }
+
+    Ok(sigbs)
```

Comment:
i *think* this should largely reduce to

```rust
    sigmessage.signatures
        .find_map(|sig| {
            verify_sig_pubkey(sig, pubkeyfile)
                .map(Vec::into_boxed_slice)
                .ok()
        })
        .ok_or_else(|| Err("No signature found".into()))
```

`find_map()` could be turned into `filter_map()` with a `.collect()` into a `Vec<>` if the function needs to return multiple signatures in the future

Reply:
But that again results in a compile error:

```
error[E0599]: no method named `find_map` found for struct `Vec<Signature>` in the current scope               
   --> update-format-crau/src/bin/test.rs:120:27                                                              
    |                                                                                                         
120 |     sigmessage.signatures.find_map(|sig| verify_sig_pubkey(sig, pubkeyfile).map(Vec::into_boxed_slice...
    |                           ^^^^^^^^ method not found in `Vec<Signature>`                                 
                                                                                                              
For more information about this error, try `rustc --explain E0599`.                                           
error: could not compile `update-format-crau` (bin "test") due to previous error 
```

Reply:
yeah i thought there was something i was missing! could also see if taking a reference to the iterable works:

```rust
    &sigmessage.signatures
        .find_map(|sig| {
```

doesn't always, but sometimes especially in a `for` loop that'll work.

Reply:
Thanks.

This change results in further compilation issues, especially a return type mismatch.

```
error[E0308]: mismatched types                                                                                
   --> update-format-crau/src/bin/test.rs:121:5                                                               
    |                                                                                                         
92  |   fn parse_signature_data(sigbytes: &[u8], pubkeyfile: &str) -> Result<Box<[u8]>, Box<dyn Error>> {     
    |                                                                 --------------------------------- expect
ed `Result<Box<[u8]>, Box<(dyn std::error::Error + 'static)>>` because of return type                         
...                                                                                                           
121 | /     sigmessage.signatures.iter().                                                                     
122 | |         find_map(|sig|                                                                                
123 | |             verify_sig_pubkey(sig, pubkeyfile).                                                       
124 | |             map(Vec::into_boxed_slice).ok()).                                                         
125 | |         ok_or_else(|| Err("No signature found"))                                                      
    | |________________________________________________^ expected `Result<Box<[u8]>, Box<...>>`, found `Result
<Box<[u8]>, ...>`                                                                                             
    |                                                                                                         
    = note: expected enum `Result<_, Box<(dyn std::error::Error + 'static)>>`                                 
               found enum `Result<_, Result<_, &str>>`
```

I could resolve the error by definining a new structure `SigError`.
Would it be the only way? It looks like overkill to me. But I cannot figure out a better way.
Pushed the experiment to my branch https://github.com/flatcar/ue-rs/tree/dongsu/sig-header-v3-sigerror . Please have a look.

Reply:
As discussed, we simply do not rely on Result return type, but `Option` instead.
Then no SigError is needed.